### PR TITLE
Gpu/restrict all the things

### PR DESCRIPTION
**Pull request by thorstenhater**
Make all pointer arguments to kernels `__restrict__` to avoid unnecessary loads.

Here is the effect on the busyring benchmark (swapped pas -> hh) with 8192 cells on a
V100 GPU (time for model-run in seconds).
```
|----------+-------|
| Baseline | After |
|----------+-------|
|    2.347 | 2.268 |
|    2.345 | 2.262 |
|    2.321 | 2.276 |
|    2.323 | 2.267 |
|    2.330 | 2.249 |
|----------+-------|
|    2.321 | 2.249 |
|----------+-------|
```

### CODE REVIEW

**Comment by thorstenhater** [comment]
bors try

File: arbor/backends/gpu/matrix_fine.cu [hunk 1]
```diff
@@ -252,7 +258,7 @@ void solve_matrix_fine(
 
 void gather(
     const fvm_value_type* from,
-    fvm_value_type* to,
+    fvm_value_type*  to,
```
Comment (noraabiakar) [inline arbor/backends/gpu/matrix_fine.cu:261]: can you undo the white space changes? they occur in a few more spots.

**Comment by noraabiakar** [comment]
Looks good! I just pointed out a few white-space changes that could be removed. Also there are new conflicts on the branch, unfortunately.

File: arbor/backends/gpu/matrix_fine.cu (hunk 1 above)
Comment (thorstenhater) [reply arbor/backends/gpu/matrix_fine.cu:261]: 👍

File: arbor/backends/gpu/matrix_fine.cu [hunk 2]
```diff
@@ -264,7 +270,7 @@ void gather(
 
 void scatter(
     const fvm_value_type* from,
-    fvm_value_type* to,
+    fvm_value_type*  to,
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/backends/gpu/matrix_fine.cu:273]: +1

File: arbor/backends/gpu/matrix_fine.cu [hunk 3]
```diff
@@ -276,8 +282,8 @@ void scatter(
 
 
 void assemble_matrix_fine(
-    fvm_value_type* d,
-    fvm_value_type* rhs,
+    fvm_value_type*  d,
+    fvm_value_type*  rhs,
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/backends/gpu/matrix_fine.cu:286]: +1

File: arbor/memory/fill.cu [hunk 4]
```diff
@@ -7,9 +7,9 @@ namespace gpu {
 
 template <typename T, typename I>
 __global__
-void fill_kernel(T* v, T value, I n) {
+void fill_kernel(T* __restrict__ const v, T value, I n) {
     auto tid = threadIdx.x + blockDim.x*blockIdx.x;
-
+
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/memory/fill.cu:12]: done

**Comment by thorstenhater** [comment]
bors try

**Pull request merged**